Add vera config entries support

vangorra opened the pull request:
## Breaking Change:

None

## Description:
Adding config entries support to vera component.

**Related issue (if applicable):** None

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11663

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [X] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

vangorra committed:
Adding vera config entries support.

vangorra committed:
Fixing lint error.

MartinHjelmare commented on homeassistant/components/vera/common.py:
```diff
@@ -0,0 +1,286 @@
+        return str(self.vera_device.vera_device_id)
+
+
+def initialize_controller(hass: HomeAssistant, config: dict) -> ControllerData:
```
Please don't include the refactor in the same PR as the config entries PR.

vangorra committed:
Applying minimal changes necessary to get config entries working.

MartinHjelmare commented on homeassistant/components/vera/common.py:
```diff
@@ -0,0 +1,122 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+ControllerData = NamedTuple(
```
This is part of support for multiple controllers.

vangorra replied:
That's in place to reduce duplicate code that accesses specific indexes in the hass.data dictionary. This named tuple guarantees strict data structures at runtime thus avoiding attempts to access data in None types.

MartinHjelmare commented on homeassistant/components/vera/common.py:
```diff
@@ -0,0 +1,122 @@
+"""Common vera code."""
```
This whole module, common.py, only seems necessary to support multiple controllers.

Please break out this change to a separate PR. It's ok to do that PR first and do this config entry PR later.

vangorra replied:
All the functions and data structures in common are used to support the creation of entities from a platform's async_setup_entry function.

MartinHjelmare replied:
Yes, they are used but they don't seem necessary. The point seems to be to prepare for multiple controllers. We should keep each PR minimal.

vangorra replied:
Further cleanup complete.

vangorra committed:
Addressing PR feedback by further reducing the scope of the change.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -51,27 +49,42 @@
-    "scene",
-]
 
+def setup(hass: HomeAssistant, base_config: dict) -> bool:
```
We need to change this to a coroutine and rename it to `async_setup`.

We can only use the async api from async context and vice versa. It's easier to have the main context async and schedule parts to the executor (sync). Although we don't seem to do any I/O in `setup` so it's an easy conversion.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -51,27 +49,42 @@
+    config = base_config.get(DOMAIN, [])
+
+    # Normalize the base url.
+    config[CONF_CONTROLLER] = config.get(CONF_CONTROLLER).rstrip("/")
```
If the above returns a list, this will error.

MartinHjelmare replied:
If the above returns `None` this will error.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -80,8 +93,9 @@ def stop_subscription(event):
+    )
 
     try:
         all_devices = controller.get_devices()
```
I/O must be scheduled on the executor thread pool with `hass.async_add_executor_job`.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -102,15 +116,37 @@ def stop_subscription(event):
+    """Unload Withings config entry."""
+    controller_data = hass.data[DOMAIN]
+
+    if not controller_data:
```
When can this happen?

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -102,15 +116,37 @@ def stop_subscription(event):
+        return True
+
+    for platform in get_configured_platforms(controller_data):
+        hass.async_create_task(
```
We should await all these tasks concurrently with `asyncio.gather`.

MartinHjelmare commented on homeassistant/components/vera/common.py:
```diff
@@ -0,0 +1,31 @@
+def get_configured_platforms(controller_data: ControllerData) -> Set[str]:
+    """Get configured platforms for a controller."""
+    platforms = []
+    for platform in controller_data.devices.keys():
```
Iterating a dict iterates the keys by default. We can remove `.keys()`.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class VeraFlowHandler(config_entries.ConfigFlow):
```
Pass the domain name as domain keyword argument and remove the decorator.

https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html#defining-your-config-flow

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,13 @@
+
+    async def async_step_import(self, config: dict):
+        """Handle a flow initialized by import."""
+        return self.async_create_entry(title=config.get(CONF_CONTROLLER), data=config)
```
What happens if the user passes an incorrect url? Normally we should check that we can connect to the device before creating the entry.

MartinHjelmare reviewed (changes requested):
We also need tests for the config flow.

vangorra committed:
Addressing PR feedback.

frenck commented:
Added the `docs-missing` label. Documentation needs to be adjusted for this change.

frenck commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -50,21 +51,23 @@
     config[CONF_CONTROLLER] = config.get(CONF_CONTROLLER).rstrip("/")
+    base_url = config.get(CONF_CONTROLLER)
+
+    controller, _ = veraApi.init_controller(base_url)
```
We shouldn't create the controller here. It should be created in `async_setup_entry` and in the config flow.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,25 @@
+        """Handle a flow initialized by import."""
+        base_url = config.get(CONF_CONTROLLER)
+
+        controller = self.hass.data[DOMAIN].controller
```
Create a new controller instead. When the user initiates a config flow, ie step user, the component will not be set up yet at this point. So we shouldn't depend on hass.data stored things in the config flow.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,25 @@
+        controller = self.hass.data[DOMAIN].controller
+
+        try:
+            controller.refresh_data()
```
Schedule this on the executor thread pool.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -51,37 +50,55 @@
-    controller, _ = veraApi.init_controller(base_url)
-    hass.data[VERA_CONTROLLER] = controller
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_subscription)
+    controller = hass.data[DOMAIN].controller  # type: veraApi.VeraController
```
Create a new controller here.

vangorra committed:
Fixing pyvera import to make it easier to patch.
Addressing PR feedback regarding creation of controller and scheduling of async config flow actions.

vangorra commented:
@frenck What sorts of documentation are needed? This change only adds config entries support, it doesn't change the config structure for the component.

springstan commented:
@vangorra I think the [smartthings integration documentation](https://www.home-assistant.io/integrations/smartthings/#configure-home-assistant) is a good example for that.
Hope this helps you :)

vangorra commented:
The vera integration already has sufficient documentation here: https://www.home-assistant.io/integrations/vera/
I don't see how this change would effect the existing documentation.

vangorra committed:
Updating code owners file.

vangorra committed:
Small fixes.

vangorra commented:
What additional changes are needed? The documentation request is not particularly clear.

cogneato commented:
You would want to add that is is available as an integration like the Samsung example provided earlier.

vangorra committed:
Adding a user config flow step.

vangorra committed:
Adding optional configs for user config flow.

vangorra committed:
Updating strings to be more clear to the user.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -1,25 +1,47 @@
 
+    async def async_step_user(self, config: dict = None):
+        """Handle user initiated flow."""
+        if self.hass.config_entries.async_entries(DOMAIN):
```
If we plan to allow more than one config entry, we should try to set a unique id for the config entry. That's a new feature.

https://github.com/home-assistant/home-assistant/blob/16199b7a52146c563de36ce5e50cd1b59dbed03c/homeassistant/config_entries.py#L819-L847

MartinHjelmare replied:
I think we can remove this check if we set a unique id in the finish step.

vangorra replied:
True, the unique check is performed in the finish step. However, this means the user as entered information only to be told you cannot do that. It's not a good user experience. This approach tells the user they cannot add another before entering information.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+        base_url = config.get(CONF_CONTROLLER)
+
+        try:
+            controller = pv.VeraController(base_url)
```
If this can't raise, please move it out of the try...except block.

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        handler.hass = hass
+        base_url = "http://127.0.0.1/"
+
+        result = await handler.async_step_user()
```
Please use the core interfaces to set up the tests. It will make the tests more robust.

```py
result = await hass.config_entries.flow.async_init(
    DOMAIN, context={"source": "user"}
)
```

https://github.com/home-assistant/home-assistant/blob/16199b7a52146c563de36ce5e50cd1b59dbed03c/homeassistant/data_entry_flow.py#L94-L96

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+        if config:
+            new_config = {
+                CONF_CONTROLLER: config.get(CONF_CONTROLLER),
+                CONF_LIGHTS: parse_int_list(config.get(CONF_LIGHTS, "")),
```
The main flow should only include required items for the connection. Optional items should go in an options flow.

https://developers.home-assistant.io/docs/en/config_entries_options_flow_handler.html

vangorra committed:
Adding options flow.
Fixing some PR feedback.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -51,42 +50,46 @@
     base_url = config.get(CONF_CONTROLLER)
-    light_ids = config.get(CONF_LIGHTS)
-    exclude_ids = config.get(CONF_EXCLUDE)
+    light_ids = config_entry.options.get(CONF_LIGHTS, [])
```
If the entry was created from config yaml the options will be in config entry data now. Maybe update the entry.options here if the options in the entry.data changes?

vangorra committed:
Better handling of options.
PR feedback changes.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -72,10 +72,14 @@
     base_url = config.get(CONF_CONTROLLER)
-    light_ids = config_entry.options.get(CONF_LIGHTS, [])
-    exclude_ids = config_entry.options.get(CONF_EXCLUDE, [])
+    light_ids = config_entry.options.setdefault(
```
We're not allowed to update the options directly. We need to use `hass.config_entries.async_update_entry` to update the options.

vangorra committed:
Using config registry to update config options.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -73,13 +74,18 @@
     config = config_entry.data
 
     # Copy the configuration.yml options into the options.
+    if CONF_LIGHTS in config or CONF_EXCLUDE in config:
```
This will overwrite the entry options with config yaml options on each restart. We might want to compare the config yaml options from last start with the current config yaml options and only overwrite on change, since the user might have customized the options after last start.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+    async def async_step_import(self, config: dict):
+        """Handle a flow initialized by import."""
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason="already_setup")
```
If we update the existing entry options here instead, we have both the existing entry data and new config yaml data available to compare and can decide if we should update and overwrite entry options or not.

vangorra replied:
I think the more consistent option would be to disable options editing when configured from configuration.yml. If they are configuring from the file, then we should always honor that and not overwrite with UI provided values at all. It looks like HA's direction for configuration is 100% UI based, so discouraging/deprecating this feature would fit. Thoughts?

MartinHjelmare replied:
If we expose options in the GUI, changes made there should not be overwritten just because the user restarts home assistant. The latest set options should apply.

You can decide if you want to expose options in config yaml, in the GUI or both. But the rule above still applies.

Having options both in yaml and GUI requires more code maintenance, so it's fair if you decide to limit it to one way.

vangorra committed:
Better managing config from file or config from UI
Disabling config through UI if config is provided from a file.
More tests to account for these adjustments.

MartinHjelmare commented on homeassistant/components/vera/__init__.py:
```diff
@@ -56,34 +57,62 @@
+                "Removing existing import config for %s",
+                entry.data.get(CONF_CONTROLLER),
+            )
+            await hass.config_entries.async_remove(entry.entry_id)
```
It's not good to remove the existing entry since users will loose their customizations.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -34,33 +60,15 @@ def __init__(self, config_entry: config_entries.ConfigEntry):
     async def async_step_init(self, user_input=None):
         """Manage the options."""
+        if self.config_entry.data.get(CONF_SOURCE) == config_entries.SOURCE_IMPORT:
+            return self.async_show_form(step_id="from_config")
```
I can't find this step.

MartinHjelmare commented on tests/components/vera/test_init.py:
```diff
@@ -42,6 +45,73 @@
+
+async def test_async_setup_remove_configs(hass: HomeAssistant) -> None:
+    """Test function."""
+    entry1 = MagicMock(spec=ConfigEntry)
```
Use `MockConfigEntry`.

https://github.com/home-assistant/home-assistant/blob/7d506bc38b73f5a81649a0fe629f054ea5dbdae6/tests/common.py#L656

MartinHjelmare commented on tests/components/vera/test_init.py:
```diff
@@ -42,6 +45,73 @@
+    entry2.domain = DOMAIN
+    entry2.data = {CONF_CONTROLLER: "url2", CONF_SOURCE: config_entries.SOURCE_USER}
+
+    hass.config_entries.async_entries = MagicMock(return_value=[entry1, entry2])
```
Don't patch the core. Keep it intact and check that the correct entry data is present after updating the entries.

vangorra replied:
I'm mocking core because I don't see a good way to add config entries before running the setup hook.

MartinHjelmare replied:
Use this:
https://github.com/home-assistant/home-assistant/blob/7d506bc38b73f5a81649a0fe629f054ea5dbdae6/tests/common.py#L692-L694

MartinHjelmare commented on tests/components/vera/test_init.py:
```diff
@@ -1,78 +1,152 @@
+    hass.config_entries.async_entries = MagicMock(return_value=[entry1, entry2])
+    hass.config_entries.async_remove = remove_mock = CoroutineMock()
+
+    await async_setup(hass, {})
```
Always use core interfaces to set up tests. Use `async_setup_component`.

vangorra committed:
Address PR feedback.

vangorra committed:
Fixing test, merging with master.

MartinHjelmare commented on homeassistant/components/vera/strings.json:
```diff
@@ -0,0 +1,39 @@
+      },
+      "from_config": {
+        "title": "UI config not supported",
+        "description": "Controller was configured using configuration.yml. Config changes should be made in the config file."
```
This is weird. Either we support GUI options all the way, or we should remove it.

vangorra committed:
Disabling all Vera UI for configs managed by configuration.yml.
Adding more tests.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -87,6 +79,9 @@ class VeraFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
     def async_get_options_flow(config_entry) -> OptionsFlowHandler:
         """Get the options flow."""
+        if config_entry.data.get(CONF_SOURCE) == config_entries.SOURCE_IMPORT:
+            raise data_entry_flow.UnknownHandler
```
We shouldn't try to come up with a way to not allow users to edit imported entries. Either fully support both GUI and yaml options or pick one or the other. Eg it's ok to only allow some options via config yaml.

vangorra replied:
If that's the case, then those options would be overwritten by the values in configuration.yml on every start.

MartinHjelmare replied:
If we only allow some options via config yaml, the user can only change them via config yaml so overwriting them on every restart is fine.

vangorra replied:
The options available in the UI and configuration file are the same. The only reason I'm keeping config file support is for backwards comparability. All future integrations should be done through the UI.

MartinHjelmare replied:
To support both the best way is this:

- Config yaml options is stored in config entry data at the end of the config flow for new entries.
- If the same entry, checked via unique_id, is imported from config yaml into config flow, we check if config yaml is the same as entry data. If it's the same we just abort. If the new data has changed compared to old data we update both entry data and entry options with the new config yaml data and then abort.
- Default options are set in entry options in `async_setyp_entry` if no options are set yet. The default options data is taken from entry data if it exist there due to config yaml imported entry, otherwise the defaults of the config schema.

vangorra committed:
Updating config based on unique_id.
Addressing additional PR feedback.

vangorra commented:
Not stale. Waiting for review.

MartinHjelmare commented on homeassistant/components/vera/config_flow.py:
```diff
@@ -0,0 +1,154 @@
+                reason="cannot_connect", description_placeholders={"base_url": base_url}
+            )
+
+        domain_entries = self.hass.config_entries.async_entries(DOMAIN)
```
Set the unique_id first. Then use `self._abort_if_unique_id_configured` and pass the updated data as `updates` parameter.

MartinHjelmare reviewed (commented):
Please rebase on latest dev branch.

vangorra committed:
Rebasing off dev.
Addressing feedback.

MartinHjelmare commented on tests/components/vera/test_init.py:
```diff
@@ -1,78 +1,108 @@
+    assert entries
+
+    for config_entry in entries:
+        await async_unload_entry(hass, config_entry)
```
Please use `hass.config_entries.async_unload` instead. We should always use the core interfaces in the tests. Assert that the entry state is unloaded.

MartinHjelmare commented on tests/components/vera/test_init.py:
```diff
@@ -1,78 +1,108 @@
+    entry = MagicMock(spec=ConfigEntry)  # type: ConfigEntry
+    entry.data = {CONF_CONTROLLER: "http://127.0.0.1"}
+
+    assert not await async_setup_entry(hass, entry)
```
Pleas use `hass.config_entries.async_setup`.

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+from tests.common import MockConfigEntry
+
+
+async def test_aync_step_user_success(hass: HomeAssistant) -> None:
```
```suggestion
async def test_async_step_user_success(hass: HomeAssistant) -> None:
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+
+async def test_aync_step_user_success(hass: HomeAssistant) -> None:
+    """Test function."""
```
```suggestion
    """Test user step success."""
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+        assert result.get("type") == RESULT_TYPE_FORM
+        assert result.get("step_id") == config_entries.SOURCE_USER
+
+        result = await hass.config_entries.flow.async_init(
```
Use `hass.config_entries.flow.async_configure` to continue a flow.

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+                CONF_EXCLUDE: "14 15",
+            },
+        )
+        assert result.get("type") == RESULT_TYPE_CREATE_ENTRY
```
We should use `dict[key]` when we expect the key to be in the dict.
```suggestion
        assert result["type"] == RESULT_TYPE_CREATE_ENTRY
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+
+async def test_async_step_user_already_configured(hass: HomeAssistant) -> None:
+    """Test function."""
```
```suggestion
    """Test user step with entry already configured."""
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+    handler = VeraFlowHandler()
+    handler.context = {}
+    handler.hass = hass
+
```
```suggestion
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+    assert result.get("reason") == "already_configured"
+
+
+async def test_aync_step_import_success(hass: HomeAssistant) -> None:
```
```suggestion
async def test_async_step_import_success(hass: HomeAssistant) -> None:
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+
+async def test_aync_step_import_success(hass: HomeAssistant) -> None:
+    """Test function."""
```
```suggestion
    """Test import step success."""
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+
+async def test_async_step_import_alredy_setup(hass: HomeAssistant) -> None:
+    """Test function."""
```
```suggestion
    """Test import step with entry already setup."""
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+    handler = VeraFlowHandler()
+    handler.context = {}
+    handler.hass = hass
+
```
```suggestion
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+
+async def test_async_step_finish_error(hass: HomeAssistant) -> None:
+    """Test function."""
```
```suggestion
    """Test finish step with error."""
```

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+        )
+        entries = hass.config_entries.async_entries(DOMAIN)
+
+        await hass.config_entries.options.async_init(
```
What are we testing here?

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+        controller.serial_number = "serial_number_1"
+        vera_controller_class_mock.return_value = controller
+
+        await hass.config_entries.flow.async_init(
```
If we just need an entry added, we can add a mock config entry.

MartinHjelmare commented on tests/components/vera/test_config_flow.py:
```diff
@@ -0,0 +1,187 @@
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "init"
+
+    result = await hass.config_entries.options.async_init(
```
Use `hass.config_entries.options.async_configure` to continue the options flow.

MartinHjelmare commented on tests/components/vera/common.py:
```diff
@@ -1,65 +1,123 @@
-        return ComponentData(controller=controller)
+        # Setup component through config flow.
+        if not controller_config.config_from_file:
+            await hass.config_entries.flow.async_init(
```
If we just want to set up the integration we can skip the config flow, add a mock config entry and use `hass.config_entries.async_setup` to set the config entry up.

MartinHjelmare commented on homeassistant/components/vera/strings.json:
```diff
@@ -0,0 +1,35 @@
+  },
+  "options": {
+    "abort": {
+      "file_based_config": "Configuration for this Vera controller is handled in configuration.yml. Please make changes there."
```
This isn't used.

vangorra committed:
Addressing PR feedback.

MartinHjelmare reviewed (approved):
Great!

mrand commented:
I'm worry, I can't determine - will the integration import existing yaml configs?  Will entity id's be the same with the integration as they were with yaml configs?  If not, and there is no way to migrate, this needs large red flags in the breaking changes notes because people with large Z-wave networks are in for a huge world of hurt.

vangorra commented:
This integration WILL import existing yaml configs. Entity ids WILL remain the same as they always have.
